dv/uvm/core_ibex/collect_results.py

```diff
@@ -4,78 +4,217 @@
 # SPDX-License-Identifier: Apache-2.0
 
 import argparse
+import collections
+import junit_xml
+import os.path
 import sys
-from typing import TextIO
+import yaml
+from test_run_result import TestRunResult, test_run_result_fields, \
+    check_test_run_result
+from typing import List, TextIO
+
+
+def parse_test_run_result(path: str) -> TestRunResult:
+    try:
+        with open(path) as yaml_file:
+            test_run_result_dict = yaml.load(yaml_file, Loader=yaml.SafeLoader)
+            loaded_fields = test_run_result_dict.keys()
+            if set(loaded_fields) != set(test_run_result_fields):
+                raise RuntimeError(f'Error loading YAML at {path}: does not '
+                        'contain the correct set of fields')
+
+            trr = TestRunResult(**test_run_result_dict)
+            try:
+                check_test_run_result(trr)
+            except AssertionError as e:
+                raise RuntimeError(f'Error loading YAML at path {path}: ' \
+                    'field types were incorrect')
+            return trr
+    except IOError as e:
+        raise RuntimeError(f'Error loading YAML at path {path}: {e}')
+    except yaml.YAMLError as e:
+        raise RuntimeError(f'Error parsing YAML at path {path}: {e}')
+
+def build_broken_test_run_result(err: str) -> TestRunResult:
+    return TestRunResult(
+            name = 'unknown',
+            idx = 0,
+            seed = 0,
+            binary = None,
+            uvm_log = None,
+            rtl_trace = None,
+            rtl_trace_csv = None,
+            iss_trace = None,
+            iss_trace_csv = None,
+            comparison_log = None,
+            passed = False,
+            failure_message = err
+    )
 
+def box_comment(line: str) -> str:
+    hr = '#' * 80
+    return hr + '\n# ' + line + '\n' + hr
 
-def parse_log(path: str) -> bool:
-    first_line = '(empty file)'
-    with open(path, 'r', encoding='UTF-8') as log:
-        for line in log:
-            first_line = line.rstrip()
-            break
+def gen_summary_line(passing_tests: List[TestRunResult], failing_tests:
+        List[TestRunResult]) -> str:
+    '''Generate a string summarising test results'''
+    total_tests = len(passing_tests) + len(failing_tests)
+    pass_pct = (len(passing_tests) / total_tests) * 100
 
-    if first_line == 'PASS':
-        return True
-    if first_line.startswith('FAIL'):
-        return False
+    return f'{pass_pct:0.2f}% PASS {len(passing_tests)} PASSED, ' \
+           f'{len(failing_tests)} FAILED'
 
-    raise RuntimeError('Strange first line ({!r})'.format(first_line))
+def gen_test_run_result_text(test_run_result: TestRunResult) -> str:
+    '''Generate a string describing a TestRunResult including the failure
+    message if the test did not pass'''
+    test_name_idx = f'{test_run_result.name}.{test_run_result.seed}'
+    test_underline = '-' * len(test_name_idx)
 
+    info_text = f'{test_name_idx}\n{test_underline}\n'
```

Comment:
Rather than faffing around with `\n`'s here, it's probably easier to gather a list and then join at the end. So this would be something like
```python3
    info_lines = [test_name_idx, test_underline]
```
then you'd call `info_lines.append(blah_blah_blah)` below. Finally, return `'\n'.join(info_text) + '\n'` (assuming you still want the trailing newline).